### PR TITLE
Feature/247/247 observability

### PR DESCRIPTION
**Pull request by Izzzu**
## What this PR changes/adds

This PR enables tracing observability data: logs, metrics, traces. It adds the application insights java agent to the Dockerfile.

Logs:
- Add JDK Monitor Logger extension
- Add logging configuration file

Metrics:
- Add extension for enabling jetty, jersey and default metrics 

Traces:
- Trace context propagation for state machine operations
- Add @WithSpan annotation to methods where the Participant changes the onboarding status.

Traces in Application Insights from local Registration Service set up:
Screenshot 2022-08-15 at 14 49 51

## Why it does that

We want to be able to monitor the application.

## Further notes

In the future we can clean up the Dockerfile and remove the application insights java agent jar from there. Instead we can create custom launcher for Registration Service in MVD to extend the Dockerfile adding the app insights java agent jar. 

Linked to PR in MVD: https://github.com/agera-edc/MinimumViableDataspace/pull/261

## Linked Issue(s)

Linked to [#247](https://github.com/agera-edc/MinimumViableDataspace/issues/247)

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java [hunk 1]
```diff
@@ -152,6 +161,11 @@ public Builder status(ParticipantStatus status) {
             return this;
         }
 
+        public Builder traceContext(Map<String, String> traceContext) {
+            participant.traceContext = traceContext;
```
Comment (algattik) [inline extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java]: should make immutable

**Review by marcgs — APPROVED** [review]

File: extensions/registration-service/src/test/java/org/eclipse/dataspaceconnector/registration/api/RegistrationServiceTest.java [hunk 2]
```diff
@@ -45,7 +46,7 @@ class RegistrationServiceTest {
     Monitor monitor = mock(Monitor.class);
     ParticipantStore participantStore = mock(ParticipantStore.class);
     DtoTransformerRegistry dtoTransformerRegistry = mock(DtoTransformerRegistry.class);
-    RegistrationService service = new RegistrationService(monitor, participantStore, dtoTransformerRegistry);
+    RegistrationService service = new RegistrationService(monitor, participantStore, dtoTransformerRegistry, new Telemetry());
```
Comment (cpeeyush) [inline extensions/registration-service/src/test/java/org/eclipse/dataspaceconnector/registration/api/RegistrationServiceTest.java]: Can we use mock same as others e.g. Telemetry telemetry = mock(Telemetry.class);

File: extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java [hunk 3]
```diff
@@ -48,6 +53,11 @@ public ParticipantStatus getStatus() {
         return status;
     }
 
+    @Override
+    public Map<String, String> getTraceContext() {
+        return unmodifiableMap(traceContext);
```
Comment (algattik) [inline extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java]: Not needed

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java [hunk 4]
```diff
@@ -53,7 +54,7 @@
 /**
  * EDC extension to boot the services used by the Authority Service.
  */
-@Requires({ PrivateKeyResolver.class, OkHttpClient.class, DidResolverRegistry.class, CredentialsVerifier.class })
+@Requires({ PrivateKeyResolver.class, OkHttpClient.class, DidResolverRegistry.class, CredentialsVerifier.class, Telemetry.class })
```
Comment (algattik) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java]: Unnecessary change

**Review by algattik — COMMENTED** [review]
Why are there no UT?

have you tested the changes actually work? could you post link or screenshots to App Insights?

**Review by cpeeyush — APPROVED** [review]

File: extensions/registration-service/src/test/java/org/eclipse/dataspaceconnector/registration/api/RegistrationServiceTest.java (hunk 2 above)
Comment (Izzzu) [reply extensions/registration-service/src/test/java/org/eclipse/dataspaceconnector/registration/api/RegistrationServiceTest.java]: We thought about it but the constructor here is used on purpose as it returns OpenTelemetry.noop() which should be used for testing.

File: extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java (hunk 3 above)
Comment (Izzzu) [reply extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java]: why it's not needed?

**Comment by Izzzu** [comment]
> Why are there no UT?
> 
> have you tested the changes actually work? could you post link or screenshots to App Insights?

yes, we tested it, there's a screenshot of traces from App Insights in the PR description.

What UTs do you have in mind?
The extension we are using is tested in EDC, so we thought we don't need to repeat the UTs in this case as it would be like testing the external library behaviour or logging

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java (hunk 4 above)
Comment (Izzzu) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java]: removed

File: extensions/registration-service/src/test/java/org/eclipse/dataspaceconnector/registration/api/RegistrationServiceTest.java (hunk 2 above)
Comment (Izzzu) [reply extensions/registration-service/src/test/java/org/eclipse/dataspaceconnector/registration/api/RegistrationServiceTest.java]: after double checking the tests pass so I changed to mock(Telemetry.class) as suggested to be consistent

File: extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java (hunk 3 above)
Comment (algattik) [reply extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java]: Because it’s already made unmodifiable in setter

File: extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java [hunk 5]
```diff
@@ -32,10 +36,11 @@
  * Dataspace participant.
  */
 @JsonDeserialize(builder = Participant.Builder.class)
-public class Participant {
+public class Participant implements TraceCarrier {
 
     private String did;
     private ParticipantStatus status = ONBOARDING_INITIATED;
+    private Map<String, String> traceContext = new HashMap<>();
```
Comment (algattik) [inline extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java]: Map.empty()

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/api/RegistrationService.java [hunk 6]
```diff
@@ -90,6 +93,7 @@ public void addParticipant(String did) {
         var participant = Participant.Builder.newInstance()
                 .did(did)
                 .status(ONBOARDING_INITIATED)
+                .traceContext(telemetry.getCurrentTraceContext())
```
Comment (algattik) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/api/RegistrationService.java:96]: Here, shouldn’t there be UT coverage that we are saving the context etc

File: extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java (hunk 3 above)
Comment (Izzzu) [reply extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java]: We were looking at the StatefulEntity in the EDC to have the reference how the tracecontext is handled there: https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/94563d2a7bc0e6a2ada4296f4ccb790adc8f173c/spi/common/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/entity/StatefulEntity.java
Maybe it should be also updated to keep the traceContext immutable after created in Builder

Comment (Izzzu) [reply extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java]: We don't have setters in Participant class.  If we create the Participant using builder but not providing the trace context then the context will be immutable. I can make it immutable at initialisation, but returning immutable map from getter seems to be safer

Comment (algattik) [reply extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java]: Since we are decorating it in the builder, we can save the second decorator here

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/api/RegistrationService.java (hunk 6 above)
Comment (Izzzu) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/api/RegistrationService.java:96]: I see, I updated a test in RegistrationServiceTest to include the traceContext

File: extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java (hunk 5 above)
Comment (Izzzu) [reply extensions/dataspace-authority-spi/src/main/java/org/eclipse/dataspaceconnector/registration/authority/model/Participant.java]: you mean Map.of() ? 
Changed here and in the getter
